server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java

```diff
@@ -21,36 +21,57 @@
 
 import org.apache.druid.timeline.DataSegment;
 
+import java.util.Collection;
 import java.util.List;
+import java.util.Spliterator;
 import java.util.concurrent.ThreadLocalRandom;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
 
 final class ReservoirSegmentSampler
 {
+  final static int SPLITERATOR_SIZE_THRESHOLD = 10;
 
   static BalancerSegmentHolder getRandomBalancerSegmentHolder(final List<ServerHolder> serverHolders)
   {
-    ServerHolder fromServerHolder = null;
+    ServerHolder fromServerHolder = getRandomElementFromList(serverHolders);
+    Collection<DataSegment> segments = fromServerHolder.getServer().getSegments();
+    Spliterator<DataSegment> chosenSpliterator = segments.spliterator();
     DataSegment proposalSegment = null;
-    int numSoFar = 0;
-
-    for (ServerHolder server : serverHolders) {
-      for (DataSegment segment : server.getServer().getSegments()) {
-        int randNum = ThreadLocalRandom.current().nextInt(numSoFar + 1);
-        // w.p. 1 / (numSoFar+1), swap out the server and segment
-        if (randNum == numSoFar) {
-          fromServerHolder = server;
-          proposalSegment = segment;
-        }
-        numSoFar++;
+
+    for (int i = 0; i < Math.log(segments.size()); i++) {
```

Comment:
what happens on an empty server?

Reply:
((please add unit tests))

Reply:
This is a natural logarithm, while you really want log2. `32 - Integer.numberOfLeadingZeros(size)` style of thing. Or `IntMath.log2()` from Guava.

Reply:
Do you mean a server which has no segments? I've added a test for this edge case.